File: scripts/check_webview_scripts.py
#!/usr/bin/env python3
"""Syntax-check the JavaScript embedded in webview HTML template literals.

The webview panels build their HTML — including a sizeable inline <script> — as
TypeScript template literals. Neither tsc nor eslint parses the JavaScript inside
those strings, so a syntax error there type-checks and bundles cleanly and only
surfaces as a blank, silently broken panel at runtime.

This script extracts each inline script block, substitutes the ${...}
interpolations with a literal, and runs `node --check` over the result.

Usage:  python3 scripts/check_webview_scripts.py
Exit code is non-zero if any block fails to parse.
"""

from __future__ import annotations

import pathlib
import re
import subprocess
import sys
import tempfile

SRC = pathlib.Path(__file__).resolve().parent.parent / "src"

SCRIPT_BLOCK = re.compile(r"<script(?:\s+nonce=\"\$\{nonce\}\")?\s*>(.*?)</script>", re.S)
# Matches ${...} with one level of nested braces, which covers the interpolations
# used in these templates (including ${JSON.stringify({...})}).
INTERPOLATION = re.compile(r"\$\{[^{}]*(?:\{[^{}]*\}[^{}]*)*\}")


def check_file(path: pathlib.Path) -> list[str]:
    """Return a list of failure messages for one source file."""
    failures: list[str] = []
    text = path.read_text(encoding="utf-8")

    for index, match in enumerate(SCRIPT_BLOCK.finditer(text)):
        js = INTERPOLATION.sub("null", match.group(1))
        if not js.strip():
            continue

        line_no = text[: match.start()].count("\n") + 1

        with tempfile.NamedTemporaryFile("w", suffix=".js", delete=False, encoding="utf-8") as handle:
            handle.write(js)
            tmp_name = handle.name

        try:
            result = subprocess.run(
                ["node", "--check", tmp_name],
                capture_output=True,
                text=True,
                check=False,
            )
            if result.returncode != 0:
                detail = (result.stderr or result.stdout or "").strip()
                failures.append(
                    f"{path.relative_to(SRC.parent)}:{line_no} (script block {index + 1})\n{detail}"
                )
        finally:
            pathlib.Path(tmp_name).unlink(missing_ok=True)

    return failures


def main() -> int:
    failures: list[str] = []
    checked = 0

    for path in sorted(SRC.glob("*.ts")):
        if not SCRIPT_BLOCK.search(path.read_text(encoding="utf-8")):
            continue
        checked += 1
        failures.extend(check_file(path))

    if failures:
        print("Embedded webview script syntax errors:\n", file=sys.stderr)
        for failure in failures:
            print(failure, file=sys.stderr)
            print("", file=sys.stderr)
        return 1

    print(f"Embedded webview scripts OK ({checked} file(s) checked).")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/check_webview_scripts.js
#!/usr/bin/env node
'use strict';

/**
 * Syntax-check JavaScript embedded in webview HTML template literals.
 *
 * TypeScript and ESLint do not parse JavaScript inside the HTML strings used by
 * the webview panels. This checker extracts each inline script, replaces host-side
 * `${...}` interpolations with a literal, and asks the current Node runtime to
 * parse the result. Keeping the checker in Node makes the normal npm build path
 * portable across Windows, macOS, Linux, and CI.
 */

const fs = require('fs');
const path = require('path');
const vm = require('vm');

const root = path.resolve(__dirname, '..');
const src = path.join(root, 'src');
const scriptBlock = /<script(?:\s+nonce="\$\{nonce\}")?\s*>([\s\S]*?)<\/script>/g;
// One nested brace level covers the interpolations currently used in templates,
// including `${JSON.stringify({...})}`.
const interpolation = /\$\{[^{}]*(?:\{[^{}]*\}[^{}]*)*\}/g;

function checkFile(filePath) {
    const failures = [];
    const sourceText = fs.readFileSync(filePath, 'utf8');
    let match;
    let blockIndex = 0;

    while ((match = scriptBlock.exec(sourceText)) !== null) {
        blockIndex += 1;
        const javascript = match[1].replace(interpolation, 'null');
        if (!javascript.trim()) {
            continue;
        }

        const lineNumber = sourceText.slice(0, match.index).split('\n').length;
        try {
            // Compiling without running the script performs the same syntax parse as
            // `node --check` without spawning another process.
            new vm.Script(javascript, {
                filename: `${path.relative(root, filePath)}#script-${blockIndex}`
            });
        } catch (error) {
            const detail = error instanceof Error ? error.stack || error.message : String(error);
            failures.push(
                `${path.relative(root, filePath)}:${lineNumber} ` +
                `(script block ${blockIndex})\n${detail}`
            );
        }
    }

    return { checked: blockIndex > 0, failures };
}

function main() {
    const failures = [];
    let checked = 0;

    for (const name of fs.readdirSync(src).sort()) {
        if (!name.endsWith('.ts')) {
            continue;
        }
        const result = checkFile(path.join(src, name));
        if (result.checked) {
            checked += 1;
        }
        failures.push(...result.failures);
    }

    if (failures.length > 0) {
        console.error('Embedded webview script syntax errors:\n');
        for (const failure of failures) {
            console.error(failure);
            console.error('');
        }
        return 1;
    }

    console.log(`Embedded webview scripts OK (${checked} file(s) checked).`);
    return 0;
}

process.exitCode = main();
